ls1intum/Artemis
# `Assessment`: Add possibility to sort complaints

### Checklist
#### General


#### Client

### Motivation and Context


A large course can accumulate more than 200 complaints on a single exercise. These are presented in a big table. To help keep a good overview, the option to sort these complaints

### Description

This PR add the options to sort the tables in the exercise assessment dashboards for manually assessed submissions, complaints and more feedback requests.
The idea in general is that tutors can now access the submissions that still need work easier (sorted to the top) and submissions that have waited longer are prioritized (again sorted to the top).
In particular the submissions are now sorted by default based on their submission date with older submissions being on top. And the option to start a new assessment is the top most row. Complaints and more feedback requests are sorted by default based on their possible action, the ones that still need assessment are at the top with the prioritization of those being submitted earlier.
All other columns can also be used for sorting by clicking on the name of the column in the header of the table.

Additionally the list of complaints is updated to also show the response time of complaints.

### Steps for Testing

Prerequisites:
- 1 instructor
- Multiple (>= 4) students
- 1 normal exercise
- 1 exam exercise with two assessment rounds

1. Create the exercises
2. Participate as the students
3. Assess some but not all submissions and complain or request more feedback as the student
4. Check that all tables are sorted correctly by default and that all other sorting options work
5. Check that the list of complaints (accessible via the statistics on the top right of the assessment dashboard) is displayed correctly

### Review Progress





#### Code Review
#### Manual Tests

### Test Coverage





sort.service.ts: 93.93% | 94.44% (before: 85.29% | 91.66%)

### Screenshots



Assessment Dashboard:
![Bildschirmfoto 2022-02-06 um 19 54 22](https://user-images.githubusercontent.com/38322605/152696857-fe8ade27-b89f-45e2-9a30-b6b1b206004e.png)
![Bildschirmfoto 2022-02-06 um 19 53 56](https://user-images.githubusercontent.com/38322605/152696859-ea0511b2-3633-4dc8-8251-705be6627e54.png)

List of complaints:
![Bildschirmfoto 2022-02-06 um 19 53 28](https://user-images.githubusercontent.com/38322605/152696861-ecc47ca1-0eb7-402d-b771-fbaeb2c0a05d.png)
